Allow custom object holder references to be registered

marvin-roesch opened the pull request:
This enables mods (in practice mostly those providing language adapters) to register their own custom `ObjectHolder` references. Hence, they can also do their own annotation scanning or similar while maintaining full compatibility with the existing Forge systems.

Implementation-wise, I specifically opted for the interface with a default (internal, Java-specific) implementation rather than *just* exposing the registration since the default really is tailored to the case of (static) fields in a class. In Kotlin, for example, one way to implement an `ObjectHolder`-like system is using delegates. With the interface, the delegate can register a custom reference and invalidate its internal cache when the reference is updated.

marvin-roesch committed:
Allow custom object holder references to be registered

marvin-roesch committed:
Remove unnecessary warning suppression

marvin-roesch committed:
Adhere to Forge naming standards

marvin-roesch committed:
Make default implementation public and allow non-static instances

marvin-roesch committed:
Fix formatting

mezz commented:
Looks fine to me, will need a more in-depth look from Lex since I haven't made much use of ObjectHolder.

LexManos commented on src/main/java/net/minecraftforge/registries/IObjectHolderRef.java:
```diff
@@ -0,0 +1,74 @@
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public static ForgeRegistry<?> getRegistryForType(Class<?> referenceType) {
```
If  this is static then its better to move it to the ObjectHolderRegistry class instead, or, i'm pretty sure this already exists else ware as for the registries as a whole.

marvin-roesch replied:
Looks like the method in question is `RegistryManager.getRegistry` (which `GameRegistry.findRegistry`) actually uses internally, I must have overlooked that. Problem with it (and the current method) is that it returns only the registry interface, which doesn't seem to have the `getDefault` and `isDummied` methods that the object holder system currently uses.

Since I don't want to introduce binary incompatibilities to the registry interface and there aren't any sensible defaults, I can see the following courses of action:
1) Only do the default and dummy check when the registry is the concrete `ForgeRegistry` class
2) Don't do the checks

I'm not that knowledgeable about the internals of the registry, so I'm not certain whether the dummy check is in fact necessary.

LexManos replied:
For now the best course would be to either use the ForgeRegistry class directly, or add a function<ResourceLocation, T> getter = rl -> isRegistered(rl) && !isDummied(rl) ? get(rl) : null

LexManos commented on src/main/java/net/minecraftforge/registries/ObjectHolderRegistry.java:
```diff
@@ -150,11 +150,11 @@ private void scanClassForFields(Map<String, String> classModIds, String classNam
 
-    private void addHolderReference(ObjectHolderRef ref)
+    public void addHolderReference(IObjectHolderRef ref)
     {
```
If we're exposing this publicly, we need some form of deduplication. Don't want people calling this 10x for the same fields...

marvin-roesch replied:
I think a generic `ClassMember` sort of value type would be good for this. That is to say a simple object consisting of a class (name) reference, a member name and some type information. The method can then simply check against a map/set of these member references.

LexManos replied:
Any update on implementing the deduplication?

LexManos commented on src/main/java/net/minecraftforge/registries/FieldObjectHolderRef.java:
```diff
@@ -22,34 +22,27 @@
  */
-@SuppressWarnings("rawtypes")
-class ObjectHolderRef
+public class FieldObjectHolderRef implements IObjectHolderRef
```
Why make this a public class if the only constructor is package private still? And why change the name/constructor arguments if they are only package private?

marvin-roesch replied:
The entire class is supposed to be package private, I suppose that IDEA simply was a bit overzealous with restricting access levels to the bare minimum.

The name is changed since this serves as a generic implementation for object holders actually backed by a field. The basic interface doesn't assume anything about the actual holder, as in languages other than Java creating an actual (annotated) field can be quite challenging.

marvin-roesch commented:
I'll see about addressing the issues pointed out by Lex. I noticed that we also have `ItemStack` holders. It might be worth generalising them as well. Maybe I can fit them within the concept of custom object holders.

LexManos commented:
Assigning mezz as this still needs updating, so he can reassign me when its ready.

marvin-roesch commented:
Closing this as cpw's new system will basically allow the same, controlled by the language service.
